### ISSUE TITLE
Building from a @TestingSerializable class to @JsonSerialzable does not run json_serializable.

### ISSUE THREAD

**Issue by wu-hui** [dart-lang/build#2824]
Greetings!

I am exploring an idea to build from a custom annotation (say @TestingSerializable) to a @JsonSerializable class, such that I can reuse all the code generated from `json_serializable`. 

I can generate `g.dart` file with `@JsonSerializable` now, but I am having issues to have `json_serializable` actually run against generated `g.dart`. 

I have a minimal reproduction repo here: https://github.com/wu-hui/dart_testing_serializable

In it, `lib/BuildTest` has a class `Value` annotated with `@TestingSerializable`. After running `flutter pub run build_runner build`, a `lib/BuildTest.g.dart` is generated, with a class annotated with `@JsonSerializable`. What I want is to have that generated class run by `json_serializable` again, so the `g.dart` file has the Json generation too.

Please advice if this is supported, and if yes, what did I miss. 

Thank you!

**Comment by jakemac53** [comment]
The problem here is that you are using the `SharedPartBuilder` class - these builders cannot see the outputs of other `SharedPartBuilder`s - the actual part file isn't created until they are all done.

You can instead use a `PartBuilder` (this will require importing a second part file). You will also need to write a valid constructor to the generated class (currently json_serializable will fail because one does not exist).

**Comment by jakemac53** [comment]
Here is a full diff of what I changed to get to that point:

```
diff --git a/lib/BuildTest.dart b/lib/BuildTest.dart
index 182bb78..d40ea79 100644
--- a/lib/BuildTest.dart
+++ b/lib/BuildTest.dart
@@ -1,6 +1,7 @@
 import 'package:testing_serializable/testing_serializable.dart';
 import 'package:json_annotation/json_annotation.dart';
 
+part 'BuildTest.testing_serializable.dart';
 part 'BuildTest.g.dart';
 
 @TestingSerializable()
diff --git a/lib/BuildTest.g.dart b/lib/BuildTest.g.dart
deleted file mode 100644
index caf6849..0000000
--- a/lib/BuildTest.g.dart
+++ /dev/null
@@ -1,20 +0,0 @@
-// GENERATED CODE - DO NOT MODIFY BY HAND
-
-part of 'BuildTest.dart';
-
-// **************************************************************************
-// DumbTestingGenerator
-// **************************************************************************
-
-@JsonSerializable(
-    explicitToJson: true) // explicitToJson is required by Testing API
-class _$ValueJsonSerializableHelper {
-  int a;
-
-  String b;
-
-  factory _$ValueJsonSerializableHelper.fromJson(Map<String, dynamic> data) =>
-      _$_$ValueJsonSerializableHelperFromJson(data);
-
-  Map<String, dynamic> toJson() => _$_$ValueJsonSerializableHelperToJson(this);
-}
diff --git a/local_deps/testing_serializable/build.yaml b/local_deps/testing_serializable/build.yaml
index ff62bcb..d766aef 100644
--- a/local_deps/testing_serializable/build.yaml
+++ b/local_deps/testing_serializable/build.yaml
@@ -4,6 +4,6 @@ builders:
     builder_factories: ["testingSerializable"]
     build_extensions: {".dart": [".g.dart"]}
     auto_apply: dependents
-    build_to: cache
+    build_to: source
     runs_before: ["json_serializable:json_serializable"]
-    applies_builders: ["source_gen|json_serializable:json_serializable|combining_builder"]
+    applies_builders: ["json_serializable:json_serializable"]
diff --git a/local_deps/testing_serializable/lib/testing_serializable.dart b/local_deps/testing_serializable/lib/testing_serializable.dart
index 721853a..405955c 100644
--- a/local_deps/testing_serializable/lib/testing_serializable.dart
+++ b/local_deps/testing_serializable/lib/testing_serializable.dart
@@ -13,14 +13,12 @@ class TestingSerializable {
 }
 
 Builder testingSerializable(BuilderOptions options) {
-  return SharedPartBuilder(
-      [DumbTestingGenerator()], 'testing_serializable');
+  return PartBuilder([DumbTestingGenerator()], '.testing_serializable.dart');
 }
 
 /// For every class annotated with TestingSerializable, generate a ${CLASS_NAME}JsonSerializableImpl
 /// class that is annotated with JsonSerializable.
-class DumbTestingGenerator
-    extends GeneratorForAnnotation<TestingSerializable> {
+class DumbTestingGenerator extends GeneratorForAnnotation<TestingSerializable> {
   @override
   generateForAnnotatedElement(
       Element element, ConstantReader annotation, BuildStep buildStep) {
@@ -42,9 +40,11 @@ class _GeneratorHelper {
   final ClassElement element;
   final ConstantReader annotation;
 
-  _GeneratorHelper(this._generator,
-      this.element,
-      this.annotation,);
+  _GeneratorHelper(
+    this._generator,
+    this.element,
+    this.annotation,
+  );
 
   String get _className {
     return '_\$${element.name}JsonSerializableHelper';
@@ -118,7 +118,6 @@ class _GeneratorHelper {
     yield createToJson();
     yield '}\n';
 
-
     // Support of snapshot <-> Model Class, support of DocumentId and ServerTimestamp
     // will be here.
     // yield* createFromTesting(accessibleFieldSet);
@@ -156,6 +155,3 @@ class CreateFactoryResult {
 
   CreateFactoryResult(this.output, this.usedFields);
 }
-
-
-
diff --git a/local_deps/testing_serializable/pubspec.lock b/local_deps/testing_serializable/pubspec.lock
index d6f2c46..cfa81a1 100644
--- a/local_deps/testing_serializable/pubspec.lock
+++ b/local_deps/testing_serializable/pubspec.lock
@@ -92,13 +92,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "7.1.0"
-  characters:
-    dependency: transitive
-    description:
-      name: characters
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "1.1.0-nullsafety.2"
   charcode:
     dependency: transitive
     description:
@@ -120,29 +113,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "0.2.0"
-  cloud_firestore:
-    dependency: "direct main"
-    description:
-      path: "packages/cloud_firestore/cloud_firestore"
-      ref: HEAD
-      resolved-ref: "7eb9e21f82a09c5cecebfc231b256545cf1ce073"
-      url: "https://github.com/FirebaseExtended/flutterfire.git"
-    source: git
-    version: "0.14.0+2"
-  cloud_firestore_platform_interface:
-    dependency: transitive
-    description:
-      name: cloud_firestore_platform_interface
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "2.0.1"
-  cloud_firestore_web:
-    dependency: transitive
-    description:
-      name: cloud_firestore_web
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "0.2.0+1"
   code_builder:
     dependency: transitive
     description:
@@ -199,34 +169,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "1.2.4"
-  firebase:
-    dependency: transitive
-    description:
-      name: firebase
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "7.3.0"
-  firebase_core:
-    dependency: transitive
-    description:
-      name: firebase_core
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "0.5.0"
-  firebase_core_platform_interface:
-    dependency: transitive
-    description:
-      name: firebase_core_platform_interface
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "2.0.0"
-  firebase_core_web:
-    dependency: transitive
-    description:
-      name: firebase_core_web
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "0.2.0"
   fixnum:
     dependency: transitive
     description:
@@ -234,16 +176,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "0.10.11"
-  flutter:
-    dependency: transitive
-    description: flutter
-    source: sdk
-    version: "0.0.0"
-  flutter_web_plugins:
-    dependency: transitive
-    description: flutter
-    source: sdk
-    version: "0.0.0"
   glob:
     dependency: transitive
     description:
@@ -307,6 +239,13 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "3.0.1"
+  json_serializable:
+    dependency: "direct dev"
+    description:
+      name: json_serializable
+      url: "https://pub.dartlang.org"
+    source: hosted
+    version: "3.4.1"
   logging:
     dependency: "direct main"
     description:
@@ -384,13 +323,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "1.9.2"
-  plugin_platform_interface:
-    dependency: transitive
-    description:
-      name: plugin_platform_interface
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "1.0.2"
   pool:
     dependency: transitive
     description:
@@ -447,11 +379,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "0.2.3"
-  sky_engine:
-    dependency: transitive
-    description: flutter
-    source: sdk
-    version: "0.0.99"
   source_gen:
     dependency: "direct main"
     description:
@@ -550,13 +477,6 @@ packages:
       url: "https://pub.dartlang.org"
     source: hosted
     version: "1.3.0-nullsafety.2"
-  vector_math:
-    dependency: transitive
-    description:
-      name: vector_math
-      url: "https://pub.dartlang.org"
-    source: hosted
-    version: "2.1.0-nullsafety.2"
   vm_service:
     dependency: transitive
     description:
@@ -593,5 +513,4 @@ packages:
     source: hosted
     version: "2.2.1"
 sdks:
-  dart: ">=2.10.0-0.0.dev <2.10.0"
-  flutter: ">=1.12.13+hotfix.5 <2.0.0"
+  dart: ">=2.10.0-0.0 <2.10.0"
```

**Comment by jakemac53** [comment]
Going to go ahead and close this issue for now
